### PR TITLE
Every repository gets adopted

### LINKED ISSUE #2364: Add an upstream to forked repositories
### Description

Currently, cloning a forked repository adds an `origin` remote linking to the fork on GitHub. However, there is no way to fetch changes from the parent, because no remote is added for the parent.

### Version

**GitHub Desktop version:** 0.7.0

**OS version:** macOS 10.12.3 (16D32)

### Steps to Reproduce

1. Fork a repo on GitHub
1. Clone the fork with GitHub Desktop
1. Run `git remote` on the command line

**Expected behavior:** Two remotes are present, with `origin` linking to the fork and `upstream` or *`username-of-parent-repo`* linking to the parent repo

**Actual behavior:** Only the `origin` remote is present

**Reproduces how often:** Always

### PR DESCRIPTION
**Pull request by joshaber**
In support of #2364.

Save GitHub repository parents, so that we can later use them to add the upstream remote.

This is a pretty big refactoring of the repositories store, along with a change to how we change (ha!) our models.

Since the Beginning of Time, our models have been immutable. When we wanted to change a property, we'd call a `with*` method which would return a new copy of the model with the changed property. Then we'd pass the changed object to a store method which would update the database, etc.

This was less than ideal for a few reasons. It meant changes took two steps: (1) change the object, (2) pass the changed object to the store. It also got confusing as to which was the source of truth: the model object we changed in memory, or the model object we get from the store? It was easy to accidentally change a model in memory without updating the store, or update the store but not used the updated model.

This PR changes it so that stores are entirely in charge of changes. Their change methods return the updated object. Models should, generally speaking, originate from stores.

/cc @desktop/core

### CODE REVIEW

**Comment by joshaber** [comment]
This is ready 🙏

File: app/src/lib/stores/repositories-store.ts [hunk 1]
```diff
@@ -180,30 +185,95 @@ export class RepositoriesStore {
...
+    let parent: GitHubRepository | null = null
+    if (gitHubRepository.parent) {
+      parent = await this.putGitHubRepository(endpoint, gitHubRepository.parent)
+    }
 
-    return foundRepository
+    const login = gitHubRepository.owner.login.toLowerCase()
+    const owner = await this.putOwner(endpoint, login)
+
+    const existingRepo = await this.db.gitHubRepositories
```
[the remark this answers is not in the document]
Comment (joshaber) [reply app/src/lib/stores/repositories-store.ts]: The owner that we're using for lookup is already unique per endpoint.

**Comment by joshaber** [comment]
🍯

File: app/src/lib/stores/repositories-store.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (niik) [reply app/src/lib/stores/repositories-store.ts]: Ah!

**Pull request merged**